### PR TITLE
Allow using DensityEstimators as Classifiers (instead of Transforms)

### PR DESCRIPTION
**Pull request by mamikonyana**
Add some constructs to Macrobase to be able to use Density Estimators (e.g. KDE, GMM, etc.) as bases for classification (instead of using percentile classifier after running those classes as transform)

### CODE REVIEW

**Comment by mamikonyana** [comment]
@edgan8 @pbailis please have a look

@edgan8  I implemented it slightly differently since I need to dump densities on a scored grid also, so I have to generate points and score those and I need access to density estimator. Feel free to suggest an alternative if you think other design would be better.

I also included a script to plot the data that is being dumped.

File: src/main/java/macrobase/analysis/stats/mixture/VariationalGMM.java [hunk 1]
```diff
@@ -17,7 +18,7 @@
 /**
  * Fit Gaussian Mixture models using Variational Bayes
  */
-public class VariationalGMM extends BatchMixtureModel {
+public class VariationalGMM extends BatchMixtureModel implements DensityEstimater {
```
Comment (pbailis) [inline src/main/java/macrobase/analysis/stats/mixture/VariationalGMM.java]: Estimator

File: src/main/java/macrobase/analysis/transform/BatchScoreFeatureTransform.java [hunk 2]
```diff
@@ -12,20 +12,25 @@
 public class BatchScoreFeatureTransform extends BatchTransform {
     protected BatchTrainScore batchTrainScore;
     protected MacroBaseConf conf;
+    protected boolean identityTransform;
 
     public BatchScoreFeatureTransform(MacroBaseConf conf, Iterator<Datum> input, MacroBaseConf.TransformType transformType)
             throws ConfigurationException {
         super(input);
-        this.batchTrainScore = conf.constructTransform(transformType);
+        if (transformType == MacroBaseConf.TransformType.IDENTITY) {
```
Comment (pbailis) [inline src/main/java/macrobase/analysis/transform/BatchScoreFeatureTransform.java]: Why is this necessary???

File: src/main/java/macrobase/conf/MacroBaseConf.java [hunk 3]
```diff
@@ -299,6 +345,20 @@ public Boolean getBoolean(String key, Boolean defaultValue) {
...
+            return MacroBaseDefaults.CLASSIFIER_TYPE;
+        }
+        return ClassifierType.valueOf(_conf.get(CLASSIFIER_TYPE));
+    }
+
+    public DensityEstimaterType getDensityEstimatorType() {
+        if (!_conf.containsKey(DENSITY_ESTIMATER_TYPE)) {
+            return MacroBaseDefaults.DENSITY_ESTIMATER_TYPE;
+        }
+        return DensityEstimaterType.valueOf(_conf.get(DENSITY_ESTIMATER_TYPE));
```
Comment (pbailis) [inline src/main/java/macrobase/conf/MacroBaseConf.java]: Estimator, ESTIMATOR

**Comment by edgan8** [comment]
Hmm, maybe I am missing something. It looks like you are still using percentile based cutoffs on the density scores, so what do we gain by having a distinct density classifier? It seems nicer to always use the same common percentile classifier step.

**Comment by pbailis** [comment]
+1; I strongly believe we should have only `FeatureTransform` and `OutlierClassifier` top level classes to handle this functionality.

**Comment by pbailis** [comment]
I can see some scenarios where combining the density estimation and the percentile classification would be beneficial, but I don't see any of them in this PR.
